eng/code-quality-reports/src/main/java/com/azure/tools/checkstyles/checks/ServiceClientChecks.java

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.tools.checkstyles.checks;
+
+import com.azure.common.ServiceClient;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.FullIdent;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+import java.util.Optional;
+
+/**
+ * Verifies that subclasses of {@link ServiceClient} meet a set of guidelines.
+ * <ol>
+ *  <li>They cannot have public or protected constructors</li>
+ *  <li>They must implement a public static method named builder</li>
+ * </ol>
+ */
+public class ServiceClientChecks extends AbstractCheck {
+    private static final String BUILDER_METHOD_NAME = "builder";
+
+    private static final String CONSTRUCTOR_ERROR_MESSAGE = "Descendants of ServiceClient cannot have public or protected constructors.";
+    private static final String BUILDER_ERROR_MESSAGE = "Descendants of ServiceClient must have a static method named builder.";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    /**
+     * Denotes that the PACKAGE_DEF token type is the only one that should trigger visitToken to be called.
+     * This is done to prevent inner classes from accidentally firing checks.
+     * @return AST token types that will trigger visitToken to be called.
+     */
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] { TokenTypes.PACKAGE_DEF };
```

Comment:
This could be a member variable so we don't have to instantiate it every time a get is called.